### PR TITLE
Add USB MIDI support for SAMD and nRF.

### LINKED ISSUE #672: USB Midi!
Requested by JohnPark and Noe. Then we can create simple button instruments using midi.

### PR DESCRIPTION
**Pull request by tannewt**
The API should be identical to using a UART for MIDI.

Fixes #672

### CODE REVIEW

File: shared-module/usb_midi/__init__.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#include "py/runtime.h"
+#include "py/objtuple.h"
+#include "shared-bindings/usb_midi/PortIn.h"
+#include "shared-bindings/usb_midi/PortOut.h"
+#include "supervisor/memory.h"
+#include "tusb.h"
+
+supervisor_allocation* usb_midi_allocation;
+
+static inline uint16_t word_align(uint16_t size) {
```
Comment (dhalbert) [inline shared-module/usb_midi/__init__.c]: You might want to put this in `supervisor/memory.h` or a similar place so it's available to anyone allocating aligned memory. I'd call it `align32_size` or something like that.

File: tools/gen_usb_descriptor.py [hunk 2]
```diff
@@ -256,21 +270,11 @@ def strings_in_order(cls):
...
-#     bFunctionProtocol=audio.AUDIO_PROTOCOL_V1)
-
-
 descriptor_list = []
 descriptor_list.append(cdc_iad)
-# descriptor_list.append(audio_iad)
 descriptor_list.extend(cdc_interfaces)
 descriptor_list.extend(msc_interfaces)
-# descriptor_list.append(audio_control_interface)
+descriptor_list.append(audio_control_interface)
```
Comment (dhalbert) [inline tools/gen_usb_descriptor.py]: Not sure why this isn't `audio_interfaces`, not just `audio_control_interface`.

File: shared-bindings/usb_midi/PortIn.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+
+//| .. currentmodule:: usb_midi
+//|
+//| :class:`PortIn` -- receives midi commands over USB
+//| ===================================================
+//|
+//| .. class:: PortIn()
+//|
+//|   Not currently dynamically supported.
```
Comment (dhalbert) [inline shared-bindings/usb_midi/PortIn.c]: Could you say how to get the singleton, then?

File: shared-bindings/usb_midi/PortOut.c [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+
+//| .. currentmodule:: usb_midi
+//|
+//| :class:`PortOut` -- sends midi messages to a computer over USB
+//| ==============================================================
+//|
+//| .. class:: PortOut()
+//|
+//|   Not currently dynamically supported.
```
Comment (dhalbert) [inline shared-bindings/usb_midi/PortOut.c]: Could you say how to get the singleton?

**Comment by siddacious** [comment]
Huzzah! 🎹

File: tools/gen_usb_descriptor.py (hunk 2 above)
Comment (tannewt) [reply tools/gen_usb_descriptor.py]: Comment added. The control interface manages ordering between the audio and midi subinterfaces.

**Comment by tannewt** [comment]
@dhalbert This is ready for another look when you have a chance.

**Pull request merged**